docs/proposals/super-simple-discovery-api.md

```diff
@@ -0,0 +1,206 @@
+
+The user experience for joining a cluster would be something like:
+
+```
+kubeadm join --token=ae23dc.faddc87f5a5ab458 <address>
+```
+
+**Note:** This is logically a different use of the token from TLS bootstrap.  We harmonize these usages and allow the same token to play double duty.
+
+#### Implementation Flow
+
+`kubeadm` will implement the following flow:
+
+* `kubeadm` connects to the API server address specified over TLS.  As we don't yet have a root certificate to trust, this is an insecure connection and the server certificate is not validated.
+  * Implementation note: the API server doesn't have to expose a new and special insecure HTTP endpoint.
+* `kubeadm` requests a ConfigMap containing the kubeconfig file defined above.
+  * This ConfigMap exists at a well known URL: `https://<server>/api/v1/namespaces/kube-public/configmaps/cluster-info`
+  * This ConfigMap is really public.  Users don't need to authenticate to read this ConfigMap.  In fact, the client MUST NOT use a bearer token here as we don't trust this endpoint yet.
+* The API server returns the ConfigMap with the kubeconfig contents as normal
+  * An extra data item on that ConfigMap contains a JWS signature.
+* `kubeadm` verifies the JWS and can now trust the server.  Further
+  communication is simpler as the CA certificate in the kubeconfig file can be
+  trusted.
+
+
+#### NEW: Bootstrap Token Structure
+
+To first make this work, we put some structure into the token.  It has both a token identifier and the token value, separated by a dot.  Example:
+
+```
+ae23dc.faddc87f5a5ab458
+```
+
+The first part of the token is the `token-id`.  The second part is the `token-secret`.  By having a token identifier, we make it easier to specify *which* token you are talking about without sending the token itself in the clear.
+
+To support this new type of token, we will update the CSV token authenticator to recognize these types of tokens and index them by `token-id`.  Tokens that do not conform to this method will continue to work but won't be usable using the JWT based authentication mechanism defined below.
+
+#### NEW: Bootstrap Token Secrets
+
+Bootstrap tokens are stored and managed via Kubernetes secrets in the `kube-system` namespace.  They have type `bootstrap.kubernetes.io/token`.
+
+The following keys are on the secret data:
+* **token-id**. As defined above.
+* **token-secret**. As defined above.
+* **expiration**. After this time the token should be automatically deleted.
+* **usage-bootstrap-signing**. Set to `true` to indicate this token should be used for signing bootstrap configs.  If omitted or some other string, it defaults to `false`.
+
+These secrets can be named anything but it is suggested that they be named `bootstrap-token-<token-id>`.
+
+**QUESTION:** Should we also spec out now how we can use this token for TLS bootstrap.
```

Comment:
A couple of things here -- I'd love to be able to get rid of `kube-discovery`.  Adding yet another thing to build and ship and package and start doesn't improve things.  We've already bundled a bunch of stuff in to the apiserver and controller manager.

Second, there is precedence here.  The SA JWT authenticator (optionally) will read etcd to confirm that the token used is still available in a secret.  We could do something similar for these tokens. 

See:
* https://github.com/kubernetes/kubernetes/blob/master/pkg/serviceaccount/jwt.go#L295
* https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/serviceaccount/tokengetter.go#L72

I worry that since we are so early in the cycle here, setting up an external token auth system (with the webhook) would be difficult and would create reentrancy issues due to the fact it needs to read secrets.